src/parser2.rs

```diff
@@ -0,0 +1,237 @@
+        assert!(parser.endblock_tag());
+        expect_tokens(&parser, vec![
+            Token { rule: Rule::endblock_tag, start: 0, end: 20 },
+            Token { rule: Rule::identifier, start: 12, end: 17 },
+        ]);
+    }
+
+    #[test]
+    fn test_for_tag() {
+        let mut parser = Rdp::new(StringInput::new("{% for client in clients %}"));
+        assert!(parser.for_tag());
+        expect_tokens(&parser, vec![
+            Token { rule: Rule::for_tag, start: 0, end: 27 },
+            Token { rule: Rule::identifier, start: 7, end: 13 },
+            Token { rule: Rule::identifier, start: 17, end: 24 },
+        ]);
+    }
+
+    #[test]
+    fn test_endfor_tag() {
+        let mut parser = Rdp::new(StringInput::new("{% endfor %}"));
+        assert!(parser.endfor_tag());
+        expect_tokens(&parser, vec![
+            Token { rule: Rule::endfor_tag, start: 0, end: 12 },
+        ]);
+    }
+
+    #[test]
+    fn test_expression_math() {
+        let mut parser = Rdp::new(StringInput::new("1 + 2 + 3 * 9/2 + 2"));
+        assert!(parser.expression());
+        expect_tokens(&parser, vec![
+            Token { rule: Rule::add_sub, start: 0, end: 19 },
+            Token { rule: Rule::add_sub, start: 0, end: 13 },
+            Token { rule: Rule::add_sub, start: 0, end: 5 },
+            Token { rule: Rule::int, start: 0, end: 1 },
+            Token { rule: Rule::op_plus, start: 2, end: 3 },
+            Token { rule: Rule::int, start: 4, end: 5 },
+            Token { rule: Rule::op_plus, start: 6, end: 7 },
+            Token { rule: Rule::mul_div, start: 8, end: 15 },
+            Token { rule: Rule::mul_div, start: 8, end: 13 },
+            Token { rule: Rule::int, start: 8, end: 9 },
+            Token { rule: Rule::op_times, start: 10, end: 11 },
+            Token { rule: Rule::int, start: 12, end: 13 },
+            Token { rule: Rule::op_slash, start: 13, end: 14 },
+            Token { rule: Rule::int, start: 14, end: 15 },
+            Token { rule: Rule::op_plus, start: 16, end: 17 },
+            Token { rule: Rule::int, start: 18, end: 19 }
+        ]);
+    }
+
+    #[test]
+    fn test_expression_identifier_logic_simple() {
+        let mut parser = Rdp::new(StringInput::new("index + 1 > 1"));
+        assert!(parser.expression());
+        expect_tokens(&parser, vec![
+            Token { rule: Rule::comparison, start: 0, end: 13 },
+            Token { rule: Rule::add_sub, start: 0, end: 9 },
+            Token { rule: Rule::identifier, start: 0, end: 5 },
+            Token { rule: Rule::op_plus, start: 6, end: 7 },
+            Token { rule: Rule::int, start: 8, end: 9 },
+            Token { rule: Rule::op_gt, start: 10, end: 11 },
+            Token { rule: Rule::int, start: 12, end: 13 }
+        ]);
+    }
+
+    #[test]
+    fn test_expression_identifier_logic_complex() {
+        let mut parser = Rdp::new(StringInput::new("1 > 2 || 3 == 4 && admin"));
```

Comment:
well that's a stupid mistake haha, forgot I'm switching from `|| &&` to `or and` >_<, it works now

Does that mean it will panic if it finds an unexpected char instead of erroring? Ideally it should have an error token that I can improve later on (ie, you used `||` but Tera uses `or` so replace `||` by `or` on line 23, col 19)